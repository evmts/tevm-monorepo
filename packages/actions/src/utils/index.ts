export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
