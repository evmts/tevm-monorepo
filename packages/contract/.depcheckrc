ignores: [
  "@evmts/core",
  "@evmts/ts-plugin",
  "zod"
]
skip-missing: true
