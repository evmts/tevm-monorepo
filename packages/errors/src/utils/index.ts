export type { DecodeFunctionDataError } from './DecodeFunctionDataError.js'
export type { EncodeFunctionReturnDataError } from './EncodeFunctionReturnDataError.js'
