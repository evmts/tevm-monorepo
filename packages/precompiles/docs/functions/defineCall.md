**@tevm/precompiles** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > defineCall

# Function: defineCall()

> **defineCall**\<`TAbi`\>(`abi`, `handlers`): (`__namedParameters`) => `Promise`\<`ExecResult`\>

## Type parameters

▪ **TAbi** extends `Abi`

## Parameters

▪ **abi**: `TAbi`

▪ **handlers**: `{ [TFunctionName in string]: Handler<TAbi, TFunctionName> }`

## Returns

> > (`__namedParameters`): `Promise`\<`ExecResult`\>
>
> ### Parameters
>
> ▪ **\_\_namedParameters**: `object`
>
> ▪ **\_\_namedParameters.data**: \`0x${string}\`
>
> ▪ **\_\_namedParameters.gasLimit**: `bigint`
>
> ### Source
>
> [precompiles/src/defineCall.ts:27](https://github.com/evmts/tevm-monorepo/blob/main/packages/precompiles/src/defineCall.ts#L27)
>

## Source

[precompiles/src/defineCall.ts:21](https://github.com/evmts/tevm-monorepo/blob/main/packages/precompiles/src/defineCall.ts#L21)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
