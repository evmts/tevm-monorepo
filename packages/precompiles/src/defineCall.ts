import type { CallResult } from './CallResult.js'
import { EVMErrorMessage, EvmError, type ExecResult } from '@ethereumjs/evm'
import {
	type Abi,
	type AbiEvent,
	type AbiParametersToPrimitiveTypes,
	type ExtractAbiFunction,
	type ExtractAbiFunctionNames,
	decodeFunctionData,
	encodeAbiParameters,
	encodeEventTopics,
	encodeFunctionResult,
	hexToBytes,
} from '@tevm/utils'

type Handler<
	TAbi extends Abi,
	TFunctionName extends ExtractAbiFunctionNames<TAbi>,
> = (params: {
	gasLimit: bigint
	args: AbiParametersToPrimitiveTypes<
		ExtractAbiFunction<TAbi, TFunctionName>['inputs']
	>
}) => Promise<CallResult<TAbi, TFunctionName>>

export const defineCall = <TAbi extends Abi>(
	abi: TAbi,
	handlers: {
		[TFunctionName in ExtractAbiFunctionNames<TAbi>]: Handler<
			TAbi,
			TFunctionName
		>
	},
) => {
	return async ({
		data,
		gasLimit,
	}: { data: `0x${string}`; gasLimit: bigint }): Promise<ExecResult> => {
		const d = decodeFunctionData({
			abi: abi,
			data: data,
		})
		const handler = handlers[d.functionName]
		try {
			const {
				returnValue,
				executionGasUsed,
				logs,
				error,
				blobGasUsed,
				selfdestruct,
			} = await handler({
				gasLimit: gasLimit,
				args: d.args as any,
			})
			return {
				executionGasUsed,
				...(error ? { exeptionError: error } : {}),
				...(selfdestruct ? { selfdestruct } : {}),
				...(blobGasUsed ? { blobGasUsed } : {}),
				...(logs
					? {
							logs: logs.map((log) => {
								const topics = encodeEventTopics({
									abi,
									eventName: log.name,
									args: log.inputs,
								} as any).map((topics) => hexToBytes(topics))
								const eventItem = abi.find(
									(item) => item.type === 'event' && item.name === log.name,
								) as AbiEvent
								if (!eventItem)
									throw new Error(`Event ${log.name} not found in ABI`)
								const data = encodeAbiParameters(eventItem.inputs, log.inputs)
								return [hexToBytes(log.address), topics, hexToBytes(data)]
							}),
					  }
					: {}),
				returnValue: hexToBytes(
					encodeFunctionResult({
						abi: abi,
						functionName: d.functionName as any,
						result: returnValue as any,
					} as any),
				),
			}
		} catch (e) {
			return {
				executionGasUsed: BigInt(0),
				returnValue: Buffer.alloc(0),
				exceptionError: {
					...new EvmError(EVMErrorMessage.REVERT),
					...{
						message:
							typeof e === 'string'
								? e
								: e instanceof Error
								? e.message
								: 'unknown error',
					},
				},
			}
		}
	}
}
