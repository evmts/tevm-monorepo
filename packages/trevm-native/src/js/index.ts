import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};