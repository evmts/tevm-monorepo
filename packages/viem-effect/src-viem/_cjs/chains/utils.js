'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.formattersOptimism =
	exports.parseTransactionCelo =
	exports.serializersCelo =
	exports.serializeTransactionCelo =
	exports.formattersCelo =
		void 0
const formatters_js_1 = require('./celo/formatters.js')
Object.defineProperty(exports, 'formattersCelo', {
	enumerable: true,
	get: function () {
		return formatters_js_1.formattersCelo
	},
})
const serializers_js_1 = require('./celo/serializers.js')
Object.defineProperty(exports, 'serializeTransactionCelo', {
	enumerable: true,
	get: function () {
		return serializers_js_1.serializeTransactionCelo
	},
})
Object.defineProperty(exports, 'serializersCelo', {
	enumerable: true,
	get: function () {
		return serializers_js_1.serializersCelo
	},
})
const parsers_js_1 = require('./celo/parsers.js')
Object.defineProperty(exports, 'parseTransactionCelo', {
	enumerable: true,
	get: function () {
		return parsers_js_1.parseTransactionCelo
	},
})
const formatters_js_2 = require('./optimism/formatters.js')
Object.defineProperty(exports, 'formattersOptimism', {
	enumerable: true,
	get: function () {
		return formatters_js_2.formattersOptimism
	},
})
//# sourceMappingURL=utils.js.map
