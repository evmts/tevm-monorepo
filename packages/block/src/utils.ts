import { EthjsAddress, keccak256, toType as originalToType, toBytes } from '@tevm/utils'

/**
 * Creates a zero-filled buffer of the specified length
 * @param {number} length - The length of the buffer
 * @returns {Uint8Array} A zero-filled Uint8Array
 */
export function zeros(length: number): Uint8Array {
	return new Uint8Array(length)
}

/**
 * Creates the zero address
 * @returns {EthjsAddress} The zero address
 */
export function createZeroAddress() {
	return new EthjsAddress(new Uint8Array(20))
}

/**
 * Creates an address from a public key
 * @param {Uint8Array} publicKey - The public key
 * @returns {EthjsAddress} The address
 */
export function createAddressFromPublicKey(publicKey: Uint8Array): EthjsAddress {
	// Create address from public key by taking the last 20 bytes of keccak256 hash
	const addressBytes = keccak256(publicKey, 'bytes').slice(-20)
	return new EthjsAddress(addressBytes)
}

/**
 * Wrapper for toType that handles string inputs properly
 * @param {any} input - The input value
 * @param {number} outputType - The output type
 * @returns {any} The converted value
 */
export function safeToType(input: any, outputType: number): any {
	if (input === null || input === undefined) {
		return undefined
	}

	// Handle string inputs by converting to bytes first
	if (typeof input === 'string' && outputType === 2) {
		const bytes = toBytes(input)
		return originalToType(bytes, outputType as any)
	}

	return originalToType(input, outputType as any)
}

/**
 * Gets the v value from a signature
 * @param {any} signature - The signature object
 * @returns {bigint} The v value
 */
export function getSignatureV(signature: any): bigint {
	// Handle both old format (v) and new format (recovery)
	if (signature.v !== undefined) {
		return BigInt(signature.v)
	}
	if (signature.recovery !== undefined) {
		return BigInt(signature.recovery) + 27n
	}
	throw new Error('Invalid signature format: missing v or recovery')
}
