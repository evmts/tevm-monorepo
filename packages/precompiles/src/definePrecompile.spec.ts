import { describe, expect, it } from 'bun:test'
import { SimpleContract } from '@tevm/contract'
import { encodeFunctionData, toBytes } from '@tevm/utils'
import { defineCall } from './defineCall.js'

describe(defineCall.name, () => {
	it('should define a precompile', async () => {
		let value = 20n
		const call = defineCall(SimpleContract.abi, {
			get: async () => ({ returnValue: value, executionGasUsed: 0n }),
			set: async ({ args }) => {
				value = args[0]
				return { returnValue: undefined, executionGasUsed: 0n }
			},
		})
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.read.get()),
			}),
		).toEqual({ returnValue: toBytes(value, { size: 32 }), executionGasUsed: 0n })
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.write.set(10n)),
			}),
		).toEqual({ returnValue: new Uint8Array(), executionGasUsed: 0n })
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.read.get()),
			}),
		).toEqual({ returnValue: toBytes(10n, { size: 32 }), executionGasUsed: 0n })
	})

	it('should correctly process logs', async () => {
		let value = 0n
		const call = defineCall(SimpleContract.abi, {
			set: async ({ args }) => {
				value = args[0]
				return {
					returnValue: undefined,
					executionGasUsed: 0n,
					logs: [
						{
							eventName: 'ValueSet',
							address: '0x1234567890123456789012345678901234567890',
							args: [20n, 30n],
						},
					],
				}
			},
			get: async () => ({ returnValue: value, executionGasUsed: 0n }),
		})
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.read.get()),
			}),
		).toMatchSnapshot()
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.write.set(5n)),
			}),
		).toMatchSnapshot()
		expect(
			await call({
				gasLimit: 1n,
				data: encodeFunctionData(SimpleContract.read.get()),
			}),
		).toMatchSnapshot()
	})

	it('should handle errors in the catch block', async () => {
		const call = defineCall(SimpleContract.abi, {
			get: async () => {
				throw new Error('Test Error')
			},
			set: async () => ({ returnValue: undefined, executionGasUsed: 0n }),
		})
		const result = await call({
			gasLimit: 1n,
			data: encodeFunctionData(SimpleContract.read.get()),
		})
		expect(result.executionGasUsed).toBe(BigInt(0))
		expect(result.returnValue).toEqual(new Uint8Array())
		expect(result.exceptionError).toBeDefined()
		expect(result.exceptionError).toMatchSnapshot()
	})
})
