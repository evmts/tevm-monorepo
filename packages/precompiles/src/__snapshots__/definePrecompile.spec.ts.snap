// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineCall > should correctly process logs 1`] = `
{
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
}
`;

exports[`defineCall > should correctly process logs 2`] = `
{
  "exceptionError": {
    "error": "revert",
    "errorType": "EvmError",
    "message": "ABI encoding params/values length mismatch.
Expected length (params): 1
Given length (values): 2

Version: 2.21.1",
  },
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [],
}
`;

exports[`defineCall > should correctly process logs 3`] = `
{
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    5,
  ],
}
`;

exports[`defineCall > should handle errors in the catch block 1`] = `
{
  "error": "revert",
  "errorType": "EvmError",
  "message": "Test Error",
}
`;

exports[`defineCall should correctly process logs 1`] = `
{
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
}
`;

exports[`defineCall should correctly process logs 2`] = `
{
  "exceptionError": {
    "error": "revert",
    "errorType": "EvmError",
    "message": 
"ABI encoding params/values length mismatch.
Expected length (params): 1
Given length (values): 2

Version: 2.21.1"
,
  },
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [],
}
`;

exports[`defineCall should correctly process logs 3`] = `
{
  "executionGasUsed": 0n,
  "returnValue": Uint8Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    5,
  ],
}
`;

exports[`defineCall should handle errors in the catch block 1`] = `
{
  "error": "revert",
  "errorType": "EvmError",
  "message": "Test Error",
}
`;
